File: 백준/Platinum/11375. 열혈강호/열혈강호.cc
#include <iostream>
#include <vector>
#include <cstring>

int N, M;
std::vector<int> graph[1001];
bool is_assigned[1001];
int assigned_worker[1001];

bool dfs(int worker) {
    for (int job : graph[worker]) {
        if (is_assigned[job]) {
            continue;
        }
        is_assigned[job] = true;
        if (assigned_worker[job] == 0 || dfs(assigned_worker[job])) {
            assigned_worker[job] = worker;
            return true;
        }
    }

    return false;
}

int main() {
    std::cin >> N >> M;

    int m, job;
    for (int i = 1; i < N + 1; i++) {
        std::cin >> m;
        for (int j = 0; j < m; j++) {
            std::cin >> job;
            graph[i].push_back(job);
        }
    }

    int answer = 0;
    for (int i = 1; i < N + 1; i++) {
        std::memset(is_assigned, false, sizeof(is_assigned));
        if (dfs(i)) {
            answer++;
        }
    }

    std::cout << answer << "\n";

    return 0;
}

File: 백준/Platinum/11375. 열혈강호/열혈강호.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.List;
import java.util.StringTokenizer;

public class Main {

    static int N, M;
    static List<List<Integer>> pair = new ArrayList<>();
    static boolean[] visited;
    static int[] assignedWorker;

    public static void main(String[] args) throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        StringTokenizer stk = new StringTokenizer(br.readLine());
        N = Integer.parseInt(stk.nextToken());
        M = Integer.parseInt(stk.nextToken());

        assignedWorker = new int[M + 1];
        for (int i = 0; i < N + 1; i++) {
            pair.add(new ArrayList<>());
        }

        int works;
        for (int i = 1; i < N + 1; i++) {
            stk = new StringTokenizer(br.readLine());
            works = Integer.parseInt(stk.nextToken());
            for (int j = 0; j < works; j++) {
                pair.get(i).add(Integer.parseInt(stk.nextToken()));
            }
        }

        int answer = 0;
        for (int i = 1; i < N + 1; i++) {
            visited = new boolean[N + 1];
            if (dfs(i)) {
                answer++;
            }
        }

        System.out.println(answer);
    }

    static boolean dfs(int worker) {
        if (visited[worker]) {
            return false;
        }
        visited[worker] = true;

        for (int job : pair.get(worker)) {
            if (assignedWorker[job] == 0 || dfs(assignedWorker[job])) {
                assignedWorker[job] = worker;
                return true;
            }
        }

        return false;
    }
}